Fix Builder trying to break water

ToMe25 opened the pull request:
Closes #4796
Closes #
Closes #

# Changes proposed in this pull request:
- Fix builder trying to break fluid blocks in the clear phase
- Fix builder ignoring waterlogged blocks
- 

Review please

ToMe25 committed:
Start implementing improved water handling for the builder

ToMe25 committed:
Add some waterlogging support

LDTTeam-Buildsystem commented:
This PR has been automatically built and hosted by our build system, during this process there was a crash, error, or otherwise, please check on this at: https://teamcity.minecolonies.com/viewLog.html?buildId=38355&buildTypeId=LetSDevTogether_Minecolonies_PullRequests_2_DeployTestServer

ToMe25 committed:
Move fluid detection to its own build phase

also change fluid removal

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

ToMe25 committed:
Fix a stupid mistake

ToMe25 committed:
Finish Waterlogged block handling

ToMe25 commented:
This should be ready now.
It currently doesn't remove non-source blocks, if it should tell me, i will just copy the sponge code then.

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -960,6 +980,88 @@ protected boolean isThereAStructureToBuild()
 
+    private boolean fluidDetectStep(@NotNull final StructureIterator.StructureBlock currentBlock) {
+        worker.getCitizenStatusHandler().setLatestStatus(new TranslationTextComponent("com.minecolonies.coremod.status.detecting_fluids"));
+        if (!walkToConstructionSite(currentStructure.getCurrentBlockPosition()))
```
that needs to be properly formatted and needs javadoc, same thing for the next method.

ToMe25 replied:
Javadoc: yes, i missed that.
formatting: what is wrong with the formatting? i tried to make it like everything else.

Raycoms replied:
the bracket is on the new line

ToMe25 replied:
it was like that in clearStep so i kept it like that, or did you mean they should all get moved to a new line?

Raycoms replied:
all
{

}

Raycoms replied:
if that is in clear step then someone messed that one up (can you please put it on a new line for cleanStep too?)

ToMe25 replied:
ok

ToMe25 replied:
i hope i didn't miss any {

Raycoms replied:
awesome

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -130,6 +130,11 @@
+    /**
+     * all the fluids to be removed later.
+     */
+    private Map<Integer, List<BlockPos>> fluidsToRemove = new TreeMap<>();
```
this needs to be stored in the building class and serialized/deserialized so that it works on restart midway

ToMe25 replied:
OK i will do that.

ToMe25 replied:
uhm, in what building class?

Raycoms replied:
AbstractBuildingStructureBuilder

ToMe25 replied:
ok thanks

ToMe25 replied:
I have tried implementing this, but as i can due to the game crashing when i try to load a world with a working builder(in the dev env only) i can't test it, so i just hope it works.

Raycoms replied:
game crashing on startup of the world?

Raycoms replied:
In dev mode it should crash less not more =D

ToMe25 replied:
yes, when i load a world in the dev env and there is a builder in it the game crashes.
i did post that on discord some time in the past(with log), i just didn't make an issue for it because its only in the dev env.

Raycoms replied:
can you send this to me (via pm)?

ToMe25 replied:
yes.

ToMe25 replied:
This should work now...

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/util/StructureIterator.java:
```diff
@@ -226,11 +226,18 @@ public Result advanceBlock(final AbstractEntityAIStructure abstractEntityAIStruc
                                                                                          || structureBlock.block == Blocks.AIR
                                                                                          || !structureBlock.metadata.getMaterial().isSolid());
+            case FLUID_DETECT:
+            case FLUID_REMOVE:
```
I think we only need fluid detect, fluid remove is independent of block progress. (It's basically while list is not empty, return same state).

ToMe25 replied:
I didn't know the step would work without this, so i left it there, i will remove it then.

Raycoms replied:
you make it a normal state (like the completion state) with the if (list.hasElements ) return getState(). else return Completion (last state)

ToMe25 replied:
I did this.

ToMe25 committed:
Javadoc and formatting

ToMe25 committed:
Some Fixes

Remove Fluid_Detect build stage
Serialize/Deserialize fluidsToRemove

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingStructureBuilder.java:
```diff
@@ -79,6 +97,11 @@
+    /**
+     * all the fluids to be removed in fluids_remove.
+     */
+    private Map<Integer, List<BlockPos>> fluidsToRemove = new TreeMap<>();
```
this should be a ordered list.

ToMe25 replied:
earlier you told me to make it an ordered map, however i forgot that and made it a tree map.
what should i make it, ordered map, or list?

ToMe25 replied:
what type of map/list should this be?
java has neither a OrderedMap nor a OrderedList.

Raycoms replied:
LinkedHashMap

ToMe25 replied:
done.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingStructureBuilder.java:
```diff
@@ -408,4 +458,12 @@ public boolean requiresCompleteRequestFulfillment()
+     * Getter for the blocks to be removed in fluids_remove.
+     * @return the blocks to be removed in fluids_remove.
+     */
+    public Map<Integer, List<BlockPos>> getFluidsToRemove() {
```
bracket

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -960,6 +977,126 @@ protected boolean isThereAStructureToBuild()
+     */
+    private boolean fluidDetectStep(@NotNull final StructureIterator.StructureBlock currentBlock) {
+        worker.getCitizenStatusHandler().setLatestStatus(new TranslationTextComponent("com.minecolonies.coremod.status.detecting_fluids"));
+        if (!walkToConstructionSite(currentStructure.getCurrentBlockPosition())) {
```
bracket here and of method name

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -960,6 +977,126 @@ protected boolean isThereAStructureToBuild()
+     * @param currentBlock
+     * @return the next step once done.
+     */
+    private IAIState fluidRemoveStep() {
```
bracket

ToMe25 committed:
brackets

ToMe25 committed:
Make fluidsToRemove a LinkedHashMap

ToMe25 committed:
Remove unused import

ToMe25 committed:
remove water slower for testing purposes only

ToMe25 committed:
set the fluid removal delay back to one tick

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -158,7 +158,7 @@ protected AbstractEntityAIStructure(@NotNull final J job)
            * Check if we have to build something.
            */
-          new AITarget(IDLE, this::isThereAStructureToBuild, () -> START_BUILDING, 100),
+          new AITarget(IDLE, this::isThereAStructureToBuild, () -> START_BUILDING, 1),
```
That's the wrong AI state you reset it =D

ToMe25 replied:
I wonder why tho, i just pressed ctrl+z to undo the last change.

ToMe25 committed:
fix wrong AIState being faster

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

marchermans reviewed (dismissed):
Generally good. However even better would be to move the Constants for the nbt names to NbtConstants.

ToMe25 commented:
all from that class or only the new ones?

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

ToMe25 committed:
small changed

Move Spawn build phase to be executed last
Move NBT Tag names to NbtTagConstants

ToMe25 committed:
Make builder remove non-source blocks too

ToMe25 commented:
This should be finished now, i think.

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

ToMe25 committed:
Fix AITarget nextStep order

ToMe25 committed:
Fix switchStage not getting called

also sort the AITarget constructors

ToMe25 committed:
Fix workers taking very long to finish remove stage

ToMe25 committed:
Merge branch 'version/1.15' into fix/water

Raycoms commented:
:shipit: I'll merge it as soon as the build is finished

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/4803/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)
